src/agent/src/device.rs

```diff
@@ -379,6 +379,53 @@ pub async fn wait_for_pci_device(
     Ok(addr)
 }
 
+#[derive(Debug)]
+struct NetPciMatcher {
+    devpath: String,
+}
+
+impl NetPciMatcher {
+    fn new(relpath: &str) -> NetPciMatcher {
+        let root_bus = create_pci_root_bus_path();
+
+        NetPciMatcher {
+            devpath: format!("{}{}", root_bus, relpath),
+        }
+    }
+}
+
+impl UeventMatcher for NetPciMatcher {
+    fn is_match(&self, uev: &Uevent) -> bool {
+        uev.devpath.starts_with(self.devpath.as_str())
+            && uev.subsystem == "net"
+            && !uev.interface.is_empty()
+            && uev.action == "add"
+    }
+}
+
+pub async fn wait_for_net_interface(
+    sandbox: &Arc<Mutex<Sandbox>>,
+    pcipath: &pci::Path,
+) -> Result<()> {
+    let root_bus_sysfs = format!("{}{}", SYSFS_DIR, create_pci_root_bus_path());
+    let sysfs_rel_path = pcipath_to_sysfs(&root_bus_sysfs, pcipath)?;
+
+    let matcher = NetPciMatcher::new(&sysfs_rel_path);
+
+    // Check if the interface is already added in case network is cold-plugged
+    // or the uevent loop is started before network is added.
+    let sys_path_str = format!("{}{}", SYSFS_DIR, matcher.devpath.as_str());
+    let path = Path::new(sys_path_str.as_str());
+
+    if path.exists() {
+        return Ok(());
```

Comment:
For VFIO network device, the device is hot-plugged earlier, so the path exists at this point, this function returns OK here and `wait_for_uevent()` will not be called.  This caused update_interfaces() started to configure interface before 
```
Uevent { action: "add", devpath: "/devices/pci0000:00/0000:00:05.0/0000:02:00.0/net/eth0", devname: "", subsystem: "net", seqnum: "778", interface: "eth0" }
```
and failed with
```
error="rpc error: code = Internal desc = \"update interface: Link not found (Address: 0a:58:0a:c0:02:39)\""    
```

After comment out  `if path.exists() {...}` block, it works for VFIO device. Is this check added as a shortcut to avoid unnecessarily calling `wait_for_uevent()` in certain cases? If so, it can be removed.

Reply:
@l8huang This check is needed for virtio devices that may be added much before or are coldplugged instead of hotplugged. 
I am checking for "/sys/devices/pci0000:00/0000:00:05.0/0000:02:00.0" here, but ideally should check for path that matches something like:
"devices/pci0000:00/0000:00:05.0/0000:02:00.0/{virtio*}*/net/eth*}".
I need to test the vfio case on my side a bit more.  
Wanted to know a bit about your setup so I can replicate this on my end. The above error occurred when you passed a vfio device with sriov device plugin+k8s?

Reply:
> This check is needed for virtio devices that may be added much before or are coldplugged instead of hotplugged.

VFIO device can only be hot-plugged now. Probably this can be resolved by differentiating the check based on the interface type.

>  The above error occurred when you passed a vfio device with sriov device plugin+k8s?

yes

We have an in-house ovn-k8s which includes change for generation DAN config, that change has not been upstreamed yet. Basically I'm using a Nvidia Connect-X6 network adapter, VFIO devices are allocated from it, [sriov-network-device-plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) and ovn-k8s are used to setup pod network.

Reply:
Pushed a potential fix. Need to test this more with sriov plugin, may have to come back to this in a couple of days.

Reply:
I tested it in my environment, and it works for VFIO -- the wait_for_uevent() function was called with NetPciMatcher. 

This PR looks good to me, thanks for addressing this issue 👍

Reply:
@l8huang Thanks for testing this out.